File: day17/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"unicode"
)

const (
	puzzleInput = "input.txt"
	puzzlePart1 = 2022
	puzzlePart2 = 1_000_000_000_000
)

func main() {
	puzzleBytes, err := os.ReadFile(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	if unicode.IsSpace(rune(puzzleBytes[len(puzzleBytes)-1])) {
		puzzleBytes = puzzleBytes[:len(puzzleBytes)-1]
	}

	sim := NewSim()
	shapeCount := 0
	nextShape := 0
	nextByte := 0

	seenStates := map[int]SimState{}

	foundCycle := false
	skipped := false

	var dShape int
	var dHeight int

	skippedShapes := 0
	skippedHeight := 0

	for {
		if !sim.hasShape() {
			if shapeCount == puzzlePart1 {
				fmt.Println("Part 1:", sim.top)
			}

			if !foundCycle {
				state := nextByte*len(shapes) + nextShape
				if v, ok := seenStates[state]; ok {
					if v.count == 2 {
						dShape = shapeCount - v.shapeCount
						dHeight = sim.top - v.height
						foundCycle = true
					}
					seenStates[state] = SimState{
						count:      v.count + 1,
						shapeCount: shapeCount,
						height:     sim.top,
					}
				} else {
					seenStates[state] = SimState{
						count:      1,
						shapeCount: shapeCount,
						height:     sim.top,
					}
				}
			}

			if !skipped && foundCycle && shapeCount >= puzzlePart1 {
				cycles := (puzzlePart2 - shapeCount) / dShape
				skippedShapes = cycles * dShape
				skippedHeight = cycles * dHeight
				skipped = true
			}

			if shapeCount+skippedShapes == puzzlePart2 {
				fmt.Println("Part 2:", sim.top+skippedHeight)
				break
			}

			sim.addShape(nextShape)
			shapeCount++
			nextShape = (nextShape + 1) % len(shapes)
		}
		b := puzzleBytes[nextByte]
		nextByte = (nextByte + 1) % len(puzzleBytes)
		var pushRight bool
		switch b {
		case '<':
			pushRight = false
		case '>':
			pushRight = true
		default:
			log.Fatalln("Invalid dir")
		}
		sim.pushDir(pushRight)
		if !sim.fall() {
			sim.commitShape()
		}
	}
}

type (
	SimState struct {
		count      int
		shapeCount int
		height     int
	}

	Pos struct {
		y, x int
	}

	Sim struct {
		pos  Pos
		kind int
		grid [][]byte
		top  int
	}
)

func NewSim() *Sim {
	s := &Sim{
		pos:  Pos{},
		kind: -1,
		grid: [][]byte{},
		top:  0,
	}
	s.addRows()
	return s
}

func (s *Sim) addRows() {
	for len(s.grid) < s.top+7 {
		s.grid = append(s.grid, []byte("......."))
	}
}

func (s *Sim) hasShape() bool {
	return s.kind >= 0
}

func (s *Sim) addShape(kind int) {
	s.pos = Pos{
		y: s.top + 3,
		x: 2,
	}
	s.kind = kind
}

func (s *Sim) commitShape() {
	shape := shapes[s.kind]
	h := len(shape)
	for yp, i := range shape {
		for x, j := range i {
			if j == '#' {
				y := h - yp - 1
				ny := s.pos.y + y
				s.grid[ny][s.pos.x+x] = '#'
				if t := ny + 1; t > s.top {
					s.top = t
				}
			}
		}
	}
	s.kind = -1
	s.addRows()
}

func (s *Sim) pushDir(right bool) {
	dir := -1
	if right {
		dir = 1
	}
	if s.checkShapeCollision(0, dir) {
		return
	}
	s.pos = Pos{
		y: s.pos.y,
		x: s.pos.x + dir,
	}
}

func (s *Sim) fall() bool {
	if s.checkShapeCollision(-1, 0) {
		return false
	}
	s.pos = Pos{
		y: s.pos.y - 1,
		x: s.pos.x,
	}
	return true
}

func (s *Sim) checkShapeCollision(dy, dx int) bool {
	shape := shapes[s.kind]
	h := len(shape)
	for yp, i := range shape {
		for x, j := range i {
			if j == '#' {
				y := h - yp - 1
				if s.isGridBlock(s.pos.y+y+dy, s.pos.x+x+dx) {
					return true
				}
			}
		}
	}
	return false
}

// isShapeBlock computes if coordinate is occupied by the shape in shape
// relative coordinates. Shape data is stored in the -y direction.
func (s *Sim) isShapeBlock(kind int, y, x int) bool {
	shape := shapes[kind]
	if y < 0 || y >= len(shape) {
		return false
	}
	row := shape[len(shape)-y-1]
	if x < 0 || x >= len(row) {
		return false
	}
	return row[x] == '#'
}

// isGridBlock computes if coordinate is occupied by a barrier in grid
// coordinates. Upwards is the +y direction.
func (s *Sim) isGridBlock(y, x int) bool {
	if y < 0 || y >= len(s.grid) {
		return true
	}
	row := s.grid[y]
	if x < 0 || x >= len(row) {
		return true
	}
	return row[x] == '#'
}

func (s *Sim) render() {
	for i := len(s.grid) - 1; i >= 0; i-- {
		if s.kind >= 0 && i >= s.pos.y && i < s.pos.y+4 {
			for j := 0; j < 7; j++ {
				if s.isShapeBlock(s.kind, i-s.pos.y, j-s.pos.x) {
					fmt.Print("@")
				} else if s.isGridBlock(i, j) {
					fmt.Print("#")
				} else {
					fmt.Print(".")
				}
			}
			fmt.Println()
		} else {
			fmt.Println(string(s.grid[i]))
		}
	}
}

var (
	shapes = [][][]byte{
		{
			[]byte("####"),
		},
		{
			[]byte(".#."),
			[]byte("###"),
			[]byte(".#."),
		},
		{
			[]byte("..#"),
			[]byte("..#"),
			[]byte("###"),
		},
		{
			[]byte("#"),
			[]byte("#"),
			[]byte("#"),
			[]byte("#"),
		},
		{
			[]byte("##"),
			[]byte("##"),
		},
	}
)

File: day17/src/main.rs
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;

const PUZZLEINPUT: &str = "input.txt";
const PUZZLE_PART1: usize = 2022;
const PUZZLE_PART2: usize = 1_000_000_000_000;

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let mut file = File::open(PUZZLEINPUT)?;
    let mut puzzle_bytes = Vec::new();
    file.read_to_end(&mut puzzle_bytes)?;
    if let Some(v) = puzzle_bytes.last() {
        if v.is_ascii_whitespace() {
            puzzle_bytes.pop();
        }
    }

    let mut sim = Sim::new();
    let mut shape_count = 0;
    let mut next_shape = 0;
    let mut next_byte = 0;

    let mut seen_states = HashMap::new();

    let mut found_cycle = false;
    let mut skipped = false;

    let mut d_shape = 0;
    let mut d_height = 0;

    let mut skipped_shapes = 0;
    let mut skipped_height = 0;

    loop {
        if sim.kind == None {
            if shape_count == PUZZLE_PART1 {
                println!("Part 1: {}", sim.top);
            }

            if !found_cycle {
                let state = next_byte * SHAPES.len() + next_shape;
                match seen_states.entry(state) {
                    Entry::Occupied(mut e) => {
                        let v: &mut SimState = e.get_mut();
                        if v.count == 2 {
                            d_shape = shape_count - v.shape_count;
                            d_height = sim.top - v.height;
                            found_cycle = true;
                        }
                        *v = SimState {
                            count: v.count + 1,
                            shape_count,
                            height: sim.top,
                        }
                    }
                    Entry::Vacant(e) => {
                        e.insert(SimState {
                            count: 1,
                            shape_count,
                            height: sim.top,
                        });
                    }
                }
            }

            if !skipped && found_cycle && shape_count >= PUZZLE_PART1 {
                let cycles = (PUZZLE_PART2 - shape_count) / d_shape;
                skipped_shapes = cycles * d_shape;
                skipped_height = cycles * d_height;
                skipped = true;
            }

            if shape_count + skipped_shapes == PUZZLE_PART2 {
                println!("Part 2: {}", sim.top + skipped_height);
                break;
            }

            sim.add_shape(next_shape);
            shape_count += 1;
            next_shape = (next_shape + 1) % SHAPES.len();
        }

        let b = puzzle_bytes[next_byte];
        next_byte = (next_byte + 1) % puzzle_bytes.len();
        let push_right = match b {
            b'<' => false,
            b'>' => true,
            _ => return Err("Invalid dir".into()),
        };
        sim.push_dir(push_right);
        if !sim.fall() {
            sim.commit_shape();
        }
    }

    Ok(())
}

struct SimState {
    count: u8,
    shape_count: usize,
    height: usize,
}

struct Pos {
    y: usize,
    x: usize,
}

struct Sim {
    pos: Pos,
    kind: Option<usize>,
    grid: Vec<Vec<u8>>,
    top: usize,
}

const SHAPES: &[&[&[u8]]] = &[
    &[b"####"],
    &[b".#.", b"###", b".#."],
    &[b"..#", b"..#", b"###"],
    &[b"#", b"#", b"#", b"#"],
    &[b"##", b"##"],
];

impl Sim {
    fn new() -> Self {
        let mut s = Self {
            pos: Pos { y: 0, x: 0 },
            kind: None,
            grid: Vec::new(),
            top: 0,
        };
        s.add_rows();
        s
    }

    fn add_rows(&mut self) {
        let t = self.top + 7;
        while self.grid.len() < t {
            self.grid.push(".......".into());
        }
    }

    fn add_shape(&mut self, kind: usize) {
        self.pos = Pos {
            y: self.top + 3,
            x: 2,
        };
        self.kind = Some(kind);
    }

    fn commit_shape(&mut self) {
        let kind = match self.kind {
            Some(v) => v,
            None => return,
        };
        let shape = SHAPES[kind];
        let h = shape.len();
        for (yp, i) in shape.iter().enumerate() {
            for (x, &j) in i.iter().enumerate() {
                if j == b'#' {
                    let y = h - yp - 1;
                    let ny = self.pos.y + y;
                    self.grid[ny][self.pos.x + x] = b'#';
                    let t = ny + 1;
                    if t > self.top {
                        self.top = t;
                    }
                }
            }
        }
        self.kind = None;
        self.add_rows();
    }

    fn push_dir(&mut self, right: bool) {
        let dir = if right { 1 } else { -1 };
        if self.check_shape_collision(0, dir) {
            return;
        }
        self.pos = Pos {
            y: self.pos.y,
            x: self.pos.x.wrapping_add_signed(dir),
        };
    }

    fn fall(&mut self) -> bool {
        if self.check_shape_collision(-1, 0) {
            return false;
        }
        self.pos = Pos {
            y: self.pos.y - 1,
            x: self.pos.x,
        };
        true
    }

    fn check_shape_collision(&self, dy: isize, dx: isize) -> bool {
        let kind = match self.kind {
            Some(v) => v,
            None => return false,
        };
        let shape = SHAPES[kind];
        let h = shape.len();
        for (yp, i) in shape.iter().enumerate() {
            for (x, &j) in i.iter().enumerate() {
                if j == b'#' {
                    let y = h - yp - 1;
                    if self.is_grid_block(
                        (self.pos.y + y).wrapping_add_signed(dy),
                        (self.pos.x + x).wrapping_add_signed(dx),
                    ) {
                        return true;
                    }
                }
            }
        }
        false
    }

    fn _is_shape_block(&self, kind: usize, y: usize, x: usize) -> bool {
        let shape = SHAPES[kind];
        if y >= shape.len() {
            return false;
        }
        let row = shape[shape.len() - y - 1];
        if x >= row.len() {
            return false;
        }
        row[x] == b'#'
    }

    fn is_grid_block(&self, y: usize, x: usize) -> bool {
        if y >= self.grid.len() {
            return true;
        }
        let row = &self.grid[y];
        if x >= row.len() {
            return true;
        }
        row[x] == b'#'
    }

    fn _render(&self) {
        let mut i = self.grid.len() - 1;
        while i < self.grid.len() {
            if self.kind != None && i >= self.pos.y && i < self.pos.y + 4 {
                for j in 0..7 {
                    if self._is_shape_block(self.kind.unwrap(), i - self.pos.y, j - self.pos.x) {
                        print!("@");
                    } else if self.is_grid_block(i, j) {
                        print!("#");
                    } else {
                        print!(".");
                    }
                }
                println!();
            } else {
                println!("{}", String::from_utf8_lossy(&self.grid[i]));
            }
            i = i.wrapping_sub(1);
        }
    }
}
